pkg/resourcemanager/mysql/mysqlaaduser/reconcile.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package mysqlaaduser
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/pkg/helpers"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager/config"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager/mysql"
+	mysqldatabase "github.com/Azure/azure-service-operator/pkg/resourcemanager/mysql/database"
+
+	"github.com/Azure/azure-service-operator/api/v1alpha1"
+	"github.com/Azure/azure-service-operator/api/v1alpha2"
+	"github.com/Azure/azure-service-operator/pkg/errhelp"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager"
+
+	_ "github.com/go-sql-driver/mysql" //sql drive link
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type MySQLAADUserManager struct {
+	identityFinder *helpers.AADIdentityFinder
+	Creds          config.Credentials
+}
+
+// NewMySQLAADUserManager creates a new MySQLAADUserManager
+func NewMySQLAADUserManager(creds config.Credentials, identityFinder *helpers.AADIdentityFinder) *MySQLAADUserManager {
+	return &MySQLAADUserManager{
+		Creds:          creds,
+		identityFinder: identityFinder,
+	}
+}
+
+var _ resourcemanager.ARMClient = &MySQLAADUserManager{}
+
+// CreateUser creates aad user
+func (m *MySQLAADUserManager) CreateUser(ctx context.Context, db *sql.DB, username string, clientID string) error {
+	if err := helpers.FindBadChars(username); err != nil {
+		return fmt.Errorf("problem found with username: %v", err)
+	}
+	if err := helpers.FindBadChars(clientID); err != nil {
+		return fmt.Errorf("problem found with clientID: %v", err)
+	}
+
+	// TODO: Only need to do this once
+	// TODO: Need to talk to MySQL team to understand why we even need to do this
+	_, err := db.ExecContext(ctx, "SET aad_auth_validate_oids_in_tenant = OFF")
+	if err != nil {
+		return err
+	}
+
+	tsql := "CREATE AADUSER IF NOT EXISTS ? IDENTIFIED BY ?"
+	_, err = db.ExecContext(ctx, tsql, username, clientID)
+
+	// TODO: If we want to support arbitrary AAD users rather than just managed identity users
+	// TODO: we need to do a standard "CREATE AADUSER IF NOT EXISTS ? as ?" when clientID is empty.
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Ensure that user exists
+func (m *MySQLAADUserManager) Ensure(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	instance, err := m.convert(obj)
+	if err != nil {
+		return false, err
+	}
+
+	dbClient := mysqldatabase.GetMySQLDatabasesClient(m.Creds)
+
+	_, err = dbClient.Get(ctx, instance.Spec.ResourceGroup, instance.Spec.Server, instance.Spec.DBName)
+	if err != nil {
+		instance.Status.Message = errhelp.StripErrorIDs(err)
+
+		if mysql.IsErrorResourceNotFound(err) || mysql.IsErrorDatabaseBusy(err) {
+			return false, nil
+		}
+
+		// All other errors are fatal
+		return false, err
```

Comment:
Could use that `return false, client.IgnoreNotFound(err)` trick here (and the other similar places further down) to simplify the flow. With a different error filter function I mean.

Reply:
Actually it's not possibly for `DatabaseBusy` here anyway, so I've removed it and just turned this into `return false, mysql.IgnoreNotFound(err)`

Reply:
Fixed below too